internal/provider/kubernetes/gateway.go

```diff
@@ -280,6 +437,124 @@ func (r *gatewayReconciler) envoyServiceForGateway(ctx context.Context, gateway
 	return svc, nil
 }
 
+// gatewaysRefSecret returns true if a managed Gateway references the provided secret.
+// An error is returned if an error is encountered while checking.
+func (r *gatewayReconciler) gatewaysRefSecret(ctx context.Context, secret *corev1.Secret) (bool, error) {
+	if secret == nil {
+		return false, fmt.Errorf("secret is nil")
+	}
+	gateways := &gwapiv1b1.GatewayList{}
+	if err := r.client.List(ctx, gateways); err != nil {
+		return false, fmt.Errorf("error listing gatewayclasses: %v", err)
+	}
+	for i := range gateways.Items {
+		gw := gateways.Items[i]
+		if r.hasMatchingController(&gw) {
+			secrets, _, err := r.secretsAndRefGrantsForGateway(ctx, &gw)
+			if err != nil {
+				return false, err
+			}
+			for _, s := range secrets {
+				if s.Namespace == secret.Namespace && s.Name == secret.Name {
+					return true, nil
+				}
+			}
+		}
+	}
+
+	return false, nil
+}
+
+// secretsAndRefGrantsForGateway returns the Secrets referenced by the provided gateway listeners.
+// If the provided Gateway references a Secret in a different namespace, a list of
+// ReferenceGrants is returned that permit the cross namespace Secret reference.
+func (r *gatewayReconciler) secretsAndRefGrantsForGateway(ctx context.Context, gateway *gwapiv1b1.Gateway) ([]corev1.Secret, []gwapiv1a2.ReferenceGrant, error) {
+	var secrets []corev1.Secret
+	var returnedGrants []gwapiv1a2.ReferenceGrant
+	if terminatesTLS(gateway) {
+		for i := range gateway.Spec.Listeners {
+			listener := gateway.Spec.Listeners[i]
+			if listener.TLS != nil {
```

Comment:
`terminatesTLS` seems to loop across all listeners, and so does this line, do you think `terminatesTLS(gateway)` can be refined to `terminatesTLS(listener)` here ?